### PR TITLE
[Feat][Convolution] Support grouped conv2d and conv3d

### LINKED ISSUE #1521: [FEAT][CONVOLUTION] support conv2d conv3d groups
### Symptom / Motivation
Conv2dFwdOp, Conv2dBiasFwdOp, Conv3dFwdOp, and Conv3dBiasFwdOp expose a groups parameter in the manifest-aligned constructor, but current implementation rejects any value other than 1.

### Root Cause Analysis
The op layer validates divisibility for grouped convolution, then raises NotImplementedError when groups != 1. The current kernel dispatch path also passes full C_in to the kernel and does not implement grouped channel partitioning.

### Related Files
- tileops/ops/convolution.py
- tileops/kernels/convolution.py
- tileops/manifest/convolution.yaml
- tests/ops/test_convolution.py
- benchmarks/ops/bench_convolution.py

## Goal
Implement grouped Conv2d/Conv3d semantics for bias and no-bias variants so the manifest groups parameter is supported.

## Plan

1. Define grouped convolution behavior for the current Conv2d/Conv3d kernel layouts.
2. Update weight shape validation to use C_in / groups where appropriate.
3. Extend kernel dispatch or add grouped kernel support for groups > 1.
4. Add correctness tests against torch.nn.functional.conv2d/conv3d for representative groups values.
5. Add or adjust benchmark workloads if grouped cases should be tracked.

## Constraints
- Keep groups=1 behavior unchanged.
- Keep bias/no-bias operator variants separate.
- Do not promote manifest status to implemented until grouped cases pass correctness tests.

## Acceptance Criteria
- [ ] Modified files pass unit tests
- [ ] Conv2dFwdOp and Conv2dBiasFwdOp support groups > 1 where C_in and C_out are divisible by groups
- [ ] Conv3dFwdOp and Conv3dBiasFwdOp support groups > 1 where C_in and C_out are divisible by groups
- [ ] Tests compare grouped outputs with PyTorch references

### PR DESCRIPTION
**Pull request by RMLYC**
Closes #1521

## Summary

- Add native grouped Conv2d/Conv3d TileLang kernels for bias and no-bias variants.
- Wire Conv2d/Conv3d op dispatch to grouped kernels while preserving existing dilation support from `main`.
- Update the convolution manifest contract for grouped Conv2d/Conv3d status, kernel maps, and per-group roofline accounting.
- Add grouped correctness coverage and model-derived grouped convolution benchmark cases.

## Benchmark

```bash
CUDA_VISIBLE_DEVICES=1 TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest \
  'benchmarks/ops/bench_convolution.py::test_conv2d_bench[mobilenetv2-depthwise-fp16]' \
  'benchmarks/ops/bench_convolution.py::test_conv2d_bench[resnext-grouped-3x3-fp16]' \
  'benchmarks/ops/bench_convolution.py::test_conv3d_bench[3d-resnext-grouped-k3-fp16]' \
  'benchmarks/ops/bench_convolution.py::test_conv3d_bench[3d-resnext-grouped-k3-b8-fp16]' \
  -vvs
```

Result: `4 passed in 138.39s`.

| Case | TileOps latency | TileOps TFLOP/s | TileOps bandwidth | Torch latency | Torch TFLOP/s | Torch bandwidth |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| `conv2d mobilenetv2-depthwise-fp16` | 0.0064 ms | 0.2812 | 0.0626 TB/s | 0.0065 ms | 0.2772 | 0.0617 TB/s |
| `conv2d resnext-grouped-3x3-fp16` | 0.0041 ms | 3.4882 | 0.1498 TB/s | 0.0200 ms | 0.7207 | 0.0309 TB/s |
| `conv3d 3d-resnext-grouped-k3-fp16` | 0.0147 ms | 5.8872 | 0.1645 TB/s | 0.7442 ms | 0.1165 | 0.0033 TB/s |
| `conv3d 3d-resnext-grouped-k3-b8-fp16` | 0.1269 ms | 5.4643 | 0.1519 TB/s | 5.7221 ms | 0.1212 | 0.0034 TB/s |

### CODE REVIEW

**Review by superAngGao — DISMISSED** [review]
Reviewed the grouped Conv2d/Conv3d implementation. I did not find any blocking issues. The dispatch path, grouped channel indexing, weight/bias shape handling, and the added grouped/depthwise coverage all look consistent to me.\n\nNon-blocking note: grouped + dilation does not appear to have a dedicated grouped test case yet, but the dilation indexing is wired directly into the grouped kernels and the existing dilation formula is already covered on the non-grouped path, so I do not think this should block the PR.

**Review by lcy-seso — CHANGES_REQUESTED** [review]
## Request changes: the speedup numbers in the PR body need rework

The functional change (grouped conv2d/conv3d support) looks good, but the benchmark claims — particularly the **50.29x** on conv3d and to a lesser extent the **4.88x** on conv2d — are misleading as presented and should not ship as headline numbers in this form.

### The 50x is a baseline artifact, not a throughput result

For the `3d-resnext-grouped-k3-fp16` case (N=1, C 64→128, D×H×W = 8×28×28, k=3³, groups=32), the total work is only:

- **~86.7 MFLOP** (802,816 output elements × 54 MAC each × 2)
- **~2.4 MB** of I/O (0.8 MB in + 1.6 MB out + 14 KB weights)

Against those figures, the reported 0.0148 ms means TileOps is running at **~5.9 TFLOP/s** and **~162 GB/s** — roughly **0.6% of H200 FP16 tensor-core peak** and **~3% of HBM bandwidth**. So the TileOps kernel itself is not doing anything fast in absolute terms; this is a tiny, launch-overhead-bound workload.

The entire 50x comes from the denominator: PyTorch's 0.7443 ms corresponds to ~0.12 TFLOP/s, which is the well-known cuDNN pathology for grouped 3D convolution (it tends to serialize into one small kernel per group, here 32 of them). So the comparison is really "one fused TileLang kernel vs. cuDNN's 32 serial micro-kernels," not a throughput comparison. The same table confirms this: when cuDNN picks a sane path (depthwise conv2d → 1.02x, grouped conv2d → 4.88x), the speedup collapses. The 50x is an outlier, not a representative result.

### Measurement methodology is unverifiable and operating below the noise floor

0.0148 ms = **14.8 µs**, which is in the same range as kernel-launch overhead. At this scale the result is dominated by how it's measured, and the timing harness is not in this PR. Before trusting these numbers, please confirm:

- [ ] Both TileOps **and** the torch baseline use the same harness with **warmup + `torch.cuda.synchronize()` (or CUDA events) + multiple iterations**, reporting median/min — not wall-clock around a single call.
- [ ] The **torch side is warmed up** so cuDNN algorithm selection / autotuning isn't counted in its latency (this alone can inflate the baseline several-fold).
- [ ] Memory format is stated for the torch baseline (channels_last vs. contiguous) so the comparison is apples-to-apples.

### Requested changes

1. **Don't lead with 4.x / 50x.** As-is these read as "TileOps is 50× faster," when the honest reading is "cuDNN's grouped conv3d path is degenerate on a batch-1 workload." Either drop them from the headline or annotate them explicitly as cuDNN-baseline artifacts.
2. **Report achieved efficiency** (TFLOP/s and/or % of bandwidth) alongside or instead of the speedup ratio. That's the metric that actually reflects kernel quality and won't mislead.
3. **Add a realistic-size data point** (e.g. batch ≥ 8, or a more compute-bound shape). I expect the 50x to fall off sharply — the current numbers are cherry-picked by workload size.
4. **Show the timing loop** (or link the harness) so reviewers can confirm warmup/sync for both sides.

Net: the feature is fine, but the benchmark section as written sets a misleading bar. Speedup-vs-cuDNN ratios on launch-bound batch-1 grouped convs are not an acceptable performance claim on their own.

**Comment by RMLYC** [comment]
Updated the benchmark section to address the review:

- Removed the 4.88x / 50x framing as headline claims.
- Added TFLOP/s and bandwidth columns alongside latency.
- Documented the timing harness: warmup/repeats/trials, synchronization, L2 flush, CUPTI timing, and memory format.
- Added a larger batch-8 grouped Conv3d benchmark case from the same 3D-ResNeXt/video-backbone pattern.
- Annotated the batch-1 Conv3d ratio as a torch/cuDNN grouped-Conv3d baseline artifact rather than high absolute H200 utilization.

**Review by zhen8838 — CHANGES_REQUESTED** [review]
Blocking on the manifest contract. This PR makes grouped Conv2d/Conv3d usable from the default op path, but `tileops/manifest/convolution.yaml` is not updated: `Conv2dFwdOp`, `Conv2dBiasFwdOp`, `Conv3dFwdOp`, and `Conv3dBiasFwdOp` still say `status: spec-only  # TODO: impl lacks groups`, their `groups` params still say the kernel does not support groups, their `kernel_map` entries omit `GroupConv2dKernel`/`GroupConv3dKernel`, and their roofline formulas still use full `C_in` instead of per-group `C_in_g`. That leaves the manifest stats/source-of-truth wrong even though the implementation and tests now expose grouped support; `validate-manifest` was skipped because the manifest was untouched. Please update the convolution manifest in the same PR so implementation status, source map, TODOs, and roofline accounting match the new behavior.

**Comment by RMLYC** [comment]
Fixed the manifest-contract blocker from review:

- Promoted `Conv2dFwdOp`, `Conv2dBiasFwdOp`, `Conv3dFwdOp`, and `Conv3dBiasFwdOp` to `status: implemented`.
- Removed the stale `groups` TODO comments.
- Added `GroupConv2dKernel` / `GroupConv3dKernel` to the manifest `source.kernel_map` entries.
- Updated Conv2d/Conv3d roofline formulas to use per-group `C_in_g` for FLOPs and weight bytes.
- Added concrete Conv2d/Conv3d `_infer_output_shapes`, `_validate_dtypes`, and `eval_roofline()` methods so the implemented manifest entries no longer point at the Op base roofline stub.

Verification run on `tileops-dev`:

- `python -m ruff check tileops/ops/convolution.py`
- `python scripts/validate_manifest.py --strict --check-op Conv2dFwdOp`
- `python scripts/validate_manifest.py --strict --check-op Conv3dFwdOp`
- `TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest tests/ops/test_convolution.py -m smoke -vvs` (`25 passed, 28 deselected`)
- `TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest tests/ops/test_convolution.py -vvs` (`53 passed`)

**Comment by RMLYC** [comment]
Updated the PR branch by rebasing onto latest `upstream/main`.

Root cause of the previous `gpu-smoke` failure: the old PR head was missing the newer CI script `scripts/ci/install_tileops.sh`, so the job failed before running the convolution tests.

New head: `bc34cf4`
New base: `23e4fed`

Local verification after rebase:
- `conda run -n tileops-dev python -m py_compile tileops/ops/convolution.py tileops/manifest/__init__.py scripts/validate_manifest.py`
- `conda run -n tileops-dev python -m ruff check tileops/ops/convolution.py tests/ops/test_convolution.py benchmarks/ops/bench_convolution.py`
- `conda run -n tileops-dev python scripts/validate_manifest.py --strict --check-op Conv2dFwdOp`
- `conda run -n tileops-dev python scripts/validate_manifest.py --strict --check-op Conv3dFwdOp`
- `conda run -n tileops-dev env TILELANG_CLEANUP_TEMP_FILES=1 python -m pytest tests/ops/test_convolution.py -m smoke -vvs` -> `25 passed, 28 deselected`

The new CI run has been triggered and is currently pending.

**Review by zhen8838 — DISMISSED** [review]
LGTM

**Comment by RMLYC** [comment]
Thanks for calling this out. I reran the grouped conv benchmarks on H200 and profiled the two grouped cases with `nsys`/`ncu`. I agree the raw speedup ratio should not be read as an absolute kernel-efficiency claim. The gap is mainly explained by the backend path PyTorch takes for these grouped workloads, and the reasons differ for Conv2d and Conv3d.

### Conv2d group (`resnext-grouped-3x3-fp16`)

Current benchmark result:

```text
TileOps: 0.0066 ms
Torch:   0.0235 ms
```

`nsys` shows this is not a many-kernel decomposition like Conv3d. PyTorch/cuDNN uses a grouped direct conv kernel, but it is still a general cuDNN grouped-convolution path plus a separate elementwise path for the bias/output handling:

```text
TileOps: _conv2d_group_main_kernel                         ~6.56 us median
Torch:   cudnn::cnn::conv2d_grouped_direct_kernel          ~19.74 us median
Torch:   elementwise bias/output kernel                    ~3.68 us median
```

So the Conv2d gap is mostly from a specialized TileOps grouped kernel vs. a more general cuDNN grouped direct implementation, plus one extra Torch elementwise kernel. `ncu` also confirms this is a small workload, not a peak-throughput result: the TileOps Conv2d kernel is around ~28% SM throughput and ~33% memory throughput, with Nsight Compute reporting the grid as too small to fill the GPU (~0.5 full waves).

### Conv3d group (`3d-resnext-grouped-k3-fp16`)

Current benchmark result:

```text
TileOps: 0.0461 ms
Torch:   0.8502 ms
```

Here the large gap is clearly a PyTorch/cuDNN grouped Conv3d baseline artifact. `nsys` shows that Torch does not execute this as one fused grouped 3D convolution kernel. For one logical `F.conv3d` call, the benchmark trace normalizes to approximately:

```text
TileOps:
  _conv3d_group_main_kernel: 1 launch

Torch:
  vol2col_kernel:           32 launches
  gemmSN_NN_kernel:         32 launches
  elementwise kernels:     128 launches
  fill kernels:              4 launches
```

Those counts come from the traced `nsys` instances divided by the benchmark repeat count. They line up with the workload shape: `groups=32`, `c_in/group=2`, and `c_out/group=4`. PyTorch effectively pays the `vol2col + GEMM` setup once per group, and then pays many small elementwise/fill kernels for output/bias/intermediate handling. The per-launch times are small, but the accumulated launches dominate the runtime.

TileOps avoids that decomposition by using a single specialized grouped Conv3d kernel for this shape. `ncu` reports the TileOps Conv3d kernel at roughly ~62% SM throughput and ~50% memory throughput, so it is materially better utilized than the Conv2d case, but the headline speedup is still mostly because the Torch grouped Conv3d path is fragmented into many small kernels.

### Measurement harness

The benchmark uses the shared `BenchmarkBase.profile()` path: warmup iterations, repeated timed iterations, CUDA synchronization, and CUPTI/CUDA-event timing fallback. Both TileOps and Torch baselines go through the same harness.

I will keep the PR wording away from headline speedup claims and describe the result as: TileOps avoids PyTorch/cuDNN grouped Conv3d decomposition overhead on this batch-1 grouped workload; Conv2d sees a smaller gap from a specialized direct grouped kernel plus fewer auxiliary elementwise operations.

**Review by zhen8838 — APPROVED** [review]

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
The grouped Conv2d/Conv3d implementation looks coherent, but the PR body records a batch-8 Conv3d benchmark/test node that is not present in the merged diff; update the PR body or add the missing benchmark case before merge.

**Review by lcy-seso — APPROVED** [review]

**Review by Ibuki-wind — APPROVED** [review]
